rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,171 @@
+  individually, contributing to latency in establishing the SSH tunnel. Some of 
+  these tasks could be shared, or their results cached.
+- Monitoring a large number of short-lived `tbot` processes is difficult.
+- There is no easy way to implement global ratelimits and backoffs across all
+  the short-lived `tbot` processes. This makes gracefully handling upstream
+  outages 
+
+## Details
+
+### Implementation
+
+We will introduce a long-lived service in `tbot`: `ssh-local-proxy`.
+
+Upon start, the service will write a SSH config to a configured destination.
+This SSH config should be used by OpenSSH to leverage the service.
+
+Once initialized, the service will open a Unix socket. Upon connection to this
+socket, the client will write a short message containing the target host and
+port. The service will then establish the connection, and then begin forwarding
+data between the socket and the upstream connection.
+
+The client will be configured as the `ProxyCommand`, and additionally
+`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
+to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
+actively forwarding the connection. This means the client can exit after the
+connection to the local proxy is established.
+
+This solution means that whilst a process will be spun up for each SSH
+connection, it will return after the connection is established. This reduces the
+long-term overhead associated with each connection and the impact on the Linux
+scheduler.
+
+#### Protocol
+
+The Unix socket will implement a simple protocol:
+
+1. Upon connection, the client will send a message indicating the target.
+  a. The message will be encoded in JSON. This will allow more fields to be 
+    added in the future without breaking compatability.
+  b. The `host_port` field will be a string in the format `host:port`.
+2. The server will establish a connection to the target, and then begin
+   forwarding data between the local connection and the target.
+
+Also considered was implementing the SOCKS5 protocol. Whilst this is more 
+standardized, it is more complicated than necessary. It also relies on our
+existing implementation of a SOCKS5 server which has not received signficant
+attention in some time.
+
+#### Client Binary
+
+As it stands today, the `tbot` binary has significant overhead associated with
+it.
+
+Due to this, we will introduce a new, extremely lightweight binary that will
+act as the client.
+
+The client will accept two arguments:
+
+1. The path to the Unix socket.
+2. The message to send to the `tbot` proxy service. This will have the host and
+   port templated into it by OpenSSH with the `%h` and `%p` placeholders.
+
+Comparing Rust and Go for this lightweight binary:
+
+- Rust has a smaller binary size: 400KiB vs 1.8MiB.
+- Rust executes faster.
+- Tuning Go's GC behaviour requires also using `env` as part of the
+  ProxyCommand. This is not necessary with Rust.
```

Comment:
We could set parameters during some `init()` too, but I'm not sure if they'd cover everything and if it changes our decision, ultimately.

Reply:
Yeah - I'm not sure it makes a major difference either way but I think the very fact that we want to tune the GC is not a good sign for Go.